pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,592 @@
+	}
+}
+
+// clearPersistedAllocation ensures there is no ingress ip in the
+// service's spec and that the service's status is cleared.
+func (ic *IngressIPController) clearPersistedAllocation(service *kapi.Service) error {
+	// Assume it is safe to modify the service without worrying about changing the local cache
+
+	// An ingress ip is only allowed in ExternalIPs when a
+	// corresponding status exists, so update the spec first to avoid
+	// failing admission control.
+	ingressIP := service.Status.LoadBalancer.Ingress[0].IP
+
+	filteredIPs := service.Spec.ExternalIPs[:0]
+	changed := false
+	for _, ip := range service.Spec.ExternalIPs {
+		if ip != ingressIP {
+			filteredIPs = append(filteredIPs, ip)
+		} else {
+			changed = true
+		}
+	}
+	if changed {
+		service.Spec.ExternalIPs = filteredIPs
+		err := ic.persistSpec(service)
+		if err != nil {
+			return err
+		}
+	}
+
+	service.Status.LoadBalancer = kapi.LoadBalancerStatus{}
+	return ic.persistStatus(service)
+}
+
+// allocate assigns an unallocated ip to the given service and updates
+// the service's persisted state.
+func (ic *IngressIPController) allocate(service *kapi.Service, key string) error {
+	if !ic.syncProcessed {
+		ic.pendingAllocations = append(ic.pendingAllocations, service)
+		return nil
+	}
+
+	// Make a copy to avoid mutating cache state
+	t, err := kapi.Scheme.DeepCopy(service)
+	if err != nil {
+		return err
+	}
+	sCopy := t.(*kapi.Service)
+
+	ip, err := ic.allocateIP(sCopy.Spec.LoadBalancerIP)
+	if err != nil {
+		return err
+	}
+	ipString := ip.String()
+
+	sCopy.Status = kapi.ServiceStatus{
+		LoadBalancer: kapi.LoadBalancerStatus{
+			Ingress: []kapi.LoadBalancerIngress{
+				{
+					IP: ipString,
+				},
+			},
+		},
+	}
+	err = ic.persistStatus(sCopy)
+	if err != nil {
+		ic.release(&ip, key)
+		return err
+	}
+	ic.allocationMap[ipString] = key
+
+	return ic.ensureExternalIP(sCopy, ipString)
+}
+
+// ensureExternalIP ensures that the provided service has the ingress
+// ip persisted as an external ip.
+func (ic *IngressIPController) ensureExternalIP(service *kapi.Service, ingressIP string) error {
+	// Assume it is safe to modify the service without worrying about changing the local cache
+
+	ipExists := false
+	for _, ip := range service.Spec.ExternalIPs {
+		if ip == ingressIP {
+			ipExists = true
+			break
+		}
+	}
+	if !ipExists {
+		service.Spec.ExternalIPs = append(service.Spec.ExternalIPs, ingressIP)
+		return ic.persistSpec(service)
+	}
+	return nil
+}
+
+// allocateIP attempts to allocate the requested ip, and if that is
+// not possible, allocates the next available address.
+func (ic *IngressIPController) allocateIP(requestedIP string) (net.IP, error) {
+	var ip net.IP
+	if requestedIP != "" {
+		ip = net.ParseIP(requestedIP)
+		if ip != nil {
+			// Valid ip
+			err := ic.ipAllocator.Allocate(ip)
+			if err != nil {
+				// Requested ip cannot be allocated
+				ip = nil
+			}
+		}
+	}
+	if ip == nil {
```

Comment:
Would recommend if you have an "if" at the end of the function like this, have the last line be the line that does something relevant to the name of the function.  I.e. the method is "allocateIP", ending this with:

```
if ip != nil {
  return ip, nil
}
return ic.ipAllocator.AllocateNext()
```

would most closely follow the go/kubernetes style pattern.